server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -127,6 +144,27 @@ private TranslogWriter(
         this.seenSequenceNumbers = Assertions.ENABLED ? new HashMap<>() : null;
         this.tragedy = tragedy;
         this.operationListener = operationListener;
+        this.lastModifiedTime = new SingleObjectCache<>(TimeValue.ZERO, new LastModifiedTime(0L, -1, null)) {
+            @Override
+            protected LastModifiedTime refresh() {
+                final long lastModifiedTime;
+                try {
+                    lastModifiedTime = TranslogWriter.super.getLastModifiedTime();
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+                return new LastModifiedTime(lastModifiedTime, operationCounter, lastSyncedCheckpoint);
+            }
+
+            @Override
+            protected boolean needsRefresh() {
+                LastModifiedTime cached = getNoRefresh();
```

Comment:
Since the modification time is connected to the written file, I wonder if it would be better to judge the need for refresh based on the `getWrittenOffset()` value (which is directly got from the file channel that the translog writes to) rather than on the combination of operationCounter and lastSyncedCheckpoint?

Reply:
@kingherc thanks for your time and patient. I think it wouldn't be better if we  judge the need for refresh base on `getWrittenOffset`, because `getWrittenOffset` also invokes a syscall every time that is same as `BaseTranslogReader#getLastModifiedTime`